### PR TITLE
Add support of Xlabs Controlnets

### PR DESCRIPTION
**Pull request by Anghellia**
# What does this PR do?

Hi! 
This PR brings support of Xlabs Controlnets, so it can be used with Diffusers.
We converted checkpoints to the Diffusers format and it can be downloaded here:
- https://huggingface.co/XLabs-AI/flux-controlnet-depth-diffusers
- https://huggingface.co/XLabs-AI/flux-controlnet-canny-diffusers
- https://huggingface.co/XLabs-AI/flux-controlnet-hed-diffusers

The request: https://github.com/huggingface/diffusers/issues/9378
## Who can review?

Anyone in the community is free to review the PR once the tests have passed. 
@sayakpaul

## How to use
Here is the example of code to launch Canny Controlnet.
```
import torch
from diffusers.utils import load_image
from diffusers import FluxControlNetModel
from diffusers.pipelines import FluxControlNetPipeline
from PIL import Image
import numpy as np

generator = torch.Generator(device="cuda").manual_seed(87544357)

controlnet = FluxControlNetModel.from_pretrained(
  "Xlabs-AI/flux-controlnet-canny-diffusers",
  torch_dtype=torch.bfloat16,
  use_safetensors=True,
)
pipe = FluxControlNetPipeline.from_pretrained(
  "black-forest-labs/FLUX.1-dev",
  controlnet=controlnet,
  torch_dtype=torch.bfloat16
)
pipe.to("cuda")

control_image = load_image("https://huggingface.co/Xlabs-AI/flux-controlnet-canny-diffusers/resolve/main/canny_example.png")
prompt = "handsome girl with rainbow hair, anime"

image = pipe(
    prompt,
    control_image=control_image,
    controlnet_conditioning_scale=0.7,
    num_inference_steps=25,
    guidance_scale=3.5,
    height=1024,
    width=768,
    generator=generator,
    num_images_per_prompt=1,
).images[0]

image.save("output_test_controlnet.png")
```
## Examples

Group 1
"photo of village in the winter"

Group 2
"it programmer sitting in the office"

Group 3
"couple of man and woman in the water, dancing"

Group 4
"photo of woman in the beach"

Group 5
"futuristic bulding in the spain"

Group 6
"2d art, girl in the magic city, sparkles, fantasy"

### CODE REVIEW

**Review by a-r-r-o-w — COMMENTED** [review]
Just some minor comments that you can address/ignore based on what Sayak has to say

**Comment by sayakpaul** [comment]
@Anghellia  thanks so much for this <3

Could you supplement this PR with an example code snippet and some resultant images? Ccing @asomoza for doing a test drive, too.

File: src/diffusers/models/controlnet_flux.py
Comment (sayakpaul) [reply src/diffusers/models/controlnet_flux.py:112]: Agree.

File: src/diffusers/models/transformers/transformer_flux.py
Comment (sayakpaul) [inline src/diffusers/models/transformers/transformer_flux.py:514]:
```suggestion
                # For Xlabs ControlNet.
                if len(controlnet_block_samples) == 2:
                    hidden_states = hidden_states + controlnet_block_samples[index_block % 2]
                else:
                    hidden_states = hidden_states + controlnet_block_samples[index_block // interval_control]
```

File: src/diffusers/models/controlnet_flux.py
Comment (sayakpaul) [inline src/diffusers/models/controlnet_flux.py:59]: Is it possible to determine if a ControlNet is of type xlabs? If not, then it's fine!

**Review by sayakpaul — COMMENTED** [review]
Beautiful PR.

File: src/diffusers/models/controlnet_flux.py
Comment (Anghellia) [reply src/diffusers/models/controlnet_flux.py]: Thank you!

I updated the code with` torch.permute` and `reshape`. I also changed it to use `self.config.patch_size`, and now the patch_size is set to 2 in the model configurations on Huggingface.

File: src/diffusers/models/controlnet_flux.py
Comment (Anghellia) [reply src/diffusers/models/controlnet_flux.py:111]: I will replace it with `ControlNetConditioningEmbedding`, so we will have a standalone block.

File: src/diffusers/pipelines/flux/pipeline_flux_controlnet.py
Comment (Anghellia) [reply src/diffusers/pipelines/flux/pipeline_flux_controlnet.py]: Please check it now.

**Comment by Anghellia** [comment]
> @Anghellia thanks so much for this <3
> 
> Could you supplement this PR with an example code snippet and some resultant images? Ccing @asomoza for doing a test drive, too.

Thank you! Updated the PR with examples 🤗

File: src/diffusers/models/controlnet_flux.py
Comment (Anghellia) [reply src/diffusers/models/controlnet_flux.py:58]: Hmm, I am not sure. Actually, xlabs ControlNets are not so different from others. I see two main changes:

1. We use `num_layers=2` (the depth of ControlNet is 2). However, I don't think it's correct to rely solely on this, as one could also train a ControlNet with num_layers=2 using the diffusers script.

2. We use `input_hint_block`, but we can specify this only if we check for a specific keyword in the model's state_dict. I think this may not apply in our case.

File: src/diffusers/pipelines/flux/pipeline_flux_controlnet.py
Comment (yiyixuxu) [inline src/diffusers/pipelines/flux/pipeline_flux_controlnet.py:755]:
```suggestion
            if self.controlnet.input_hint_block is not None:
```

File: src/diffusers/models/controlnet_flux.py
Comment (yiyixuxu) [reply src/diffusers/models/controlnet_flux.py:58]: we can add a new config `conditioning_embedding_channels` to the flux controlnet that defaults to None

**Review by yiyixuxu — COMMENTED** [review]
thanks a lot for the PR! I left some suggestions, let me know if they would work!

File: src/diffusers/models/controlnet_flux.py
Comment (sayakpaul) [reply src/diffusers/models/controlnet_flux.py:58]: Maybe both could be combined to create a condition to determine if it's an Xlabs ControlNet? @yiyixuxu would love to know your thoughts here.

Comment (sayakpaul) [reply src/diffusers/models/controlnet_flux.py:58]: Nevermind I think https://github.com/huggingface/diffusers/pull/9638#discussion_r1797232049 should cut the deal for us unless there's some differences in the forward method.

**Comment by RimoChan** [comment]
I encountered an issue where the `FluxMultiControlNetModel` is not compatible with the branch I installed. I used the following command to install the package:

```
pip install git+https://github.com/XLabs-AI/diffusers.git@xlabs_controlnet_support
```

After installation, I tried executing the following code:

```python
import torch
from diffusers.utils import load_image
from diffusers import FluxControlNetModel, FluxMultiControlNetModel
from diffusers.pipelines import FluxControlNetPipeline



generator = torch.Generator(device="cuda").manual_seed(87544357)

controlnet = FluxMultiControlNetModel([
    FluxControlNetModel.from_pretrained(
        "Xlabs-AI/flux-controlnet-canny-diffusers",
        torch_dtype=torch.bfloat16,
        use_safetensors=True,
    ),
    FluxControlNetModel.from_pretrained(
        "Xlabs-AI/flux-controlnet-canny-diffusers",
        torch_dtype=torch.bfloat16,
        use_safetensors=True,
    ),
])
pipe = FluxControlNetPipeline.from_pretrained(
  '/mypath/to/FLUX.1-dev_official',
  controlnet=controlnet,
  torch_dtype=torch.bfloat16
)
pipe.to("cuda")

control_image = load_image("https://huggingface.co/Xlabs-AI/flux-controlnet-canny-diffusers/resolve/main/canny_example.png")

image = pipe(
    "handsome girl with rainbow hair, anime",
    control_image=[control_image, control_image],
    controlnet_conditioning_scale=[0.7, 0.7],
    num_inference_steps=25,
    guidance_scale=3.5,
    height=1024,
    width=768,
    generator=generator,
    num_images_per_prompt=1,
).images[0]

image.save("output_test_controlnet.png")
```

However, I encountered this error:

```
Traceback (most recent call last):
  File "/opt/tiger/test_1/t2s.py", line 33, in <module>
    image = pipe(
  File "/opt/tiger/miniconda3/lib/python3.10/site-packages/torch/utils/_contextlib.py", line 115, in decorate_context
    return func(*args, **kwargs)
  File "/opt/tiger/miniconda3/lib/python3.10/site-packages/diffusers/pipelines/flux/pipeline_flux_controlnet.py", line 897, in __call__
    controlnet_block_samples, controlnet_single_block_samples = self.controlnet(
  File "/opt/tiger/miniconda3/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1532, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
  File "/opt/tiger/miniconda3/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1541, in _call_impl
    return forward_call(*args, **kwargs)
  File "/opt/tiger/miniconda3/lib/python3.10/site-packages/diffusers/models/controlnet_flux.py", line 503, in forward
    block_samples, single_block_samples = controlnet(
  File "/opt/tiger/miniconda3/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1532, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
  File "/opt/tiger/miniconda3/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1541, in _call_impl
    return forward_call(*args, **kwargs)
  File "/opt/tiger/miniconda3/lib/python3.10/site-packages/diffusers/models/controlnet_flux.py", line 281, in forward
    controlnet_cond = self.input_hint_block(controlnet_cond)
  File "/opt/tiger/miniconda3/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1532, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
  File "/opt/tiger/miniconda3/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1541, in _call_impl
    return forward_call(*args, **kwargs)
  File "/opt/tiger/miniconda3/lib/python3.10/site-packages/diffusers/models/controlnet.py", line 99, in forward
    embedding = self.conv_in(conditioning)
  File "/opt/tiger/miniconda3/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1532, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
  File "/opt/tiger/miniconda3/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1541, in _call_impl
    return forward_call(*args, **kwargs)
  File "/opt/tiger/miniconda3/lib/python3.10/site-packages/torch/nn/modules/conv.py", line 460, in forward
    return self._conv_forward(input, self.weight, self.bias)
  File "/opt/tiger/miniconda3/lib/python3.10/site-packages/torch/nn/modules/conv.py", line 456, in _conv_forward
    return F.conv2d(input, weight, bias, self.stride,
RuntimeError: Given groups=1, weight of size [16, 3, 3, 3], expected input[1, 1, 3072, 64] to have 3 channels, but got 1 channels instead
```

Could you please investigate this incompatibility?

**Comment by sayakpaul** [comment]
MultiControlNet compatibility can be incorporated after this initial PR is merged.

File: src/diffusers/models/controlnet_flux.py
Comment (Anghellia) [reply src/diffusers/models/controlnet_flux.py]: Thanks for your suggestion, it works!

File: src/diffusers/models/transformers/transformer_flux.py
Comment (yiyixuxu) [inline src/diffusers/models/transformers/transformer_flux.py:512]: I would suggest passing a flag down here, `controlnet_repeat_interleave = False` maybe? 
This would break if someone trained a controlnet with 2 blocks but want to use the other indexing method

**Review by a-r-r-o-w — APPROVED** [review]
I think this is look good to merge now after @yiyixuxu gives a final review!

For the failing style tests, could you run `make style` and push? Thanks

**Comment by sayakpaul** [comment]
We could add some tests in a follow-up PR.

**Comment by Anghellia** [comment]
@a-r-r-o-w please launch tests

**Comment by yiyixuxu** [comment]
hey thanks for the PR! 
I merged it in here https://github.com/huggingface/diffusers/pull/9687 since I cannot push into your PR
It is branched off your PR so all your commits are there and you're an author there:)

**Comment by yiyixuxu** [comment]
closing the PR now since we already merged it!

**Pull request closed without merging**
